torchtune/utils/_checkpointing/utils.py

```diff
@@ -15,39 +15,46 @@ def update_state_dict_for_classifier(
     model_named_parameters: Iterable[Tuple[str, torch.nn.Parameter]],
 ):
     """
-    Validates the state dict for checkpoint loading for a classifier model, to be
-    used prior to a call to ``model.load_state_dict(state_dict)``.
-    When fine-tuning a classifier model from the checkpoint of a base language model
+    Validates the state dict for checkpoint loading for a classifier model.
+    To be used prior to a call to ``model.load_state_dict(state_dict)``.
+    This function will always overwrite the ``output.weight`` in the state-dict
+    to be loaded with the ``output.weight`` in the model, unless the weights
+    are already identical e.g. when resuming training.
+    For example, when fine-tuning a classifier model from the checkpoint of a base language model
     which has ``output.weight`` of shape ``[vocab_dim, embed_dim]``, we overwrite
     the ``output.weight`` in the state-dict to be loaded with the randomly initialized
     weight in the model. This is done in-place.
 
-    When fine-tuning a classifier model from the checkpoint of a base classifier model,
-    this function mostly defers to ``model.load_state_dict`` for validation, thus it is
-    reccomended to call this when ``not self._resume_from_checkpoint`` in a recipe.
-
     Args:
         state_dict (Dict[str, torch.Tensor]): state dict to be loaded into the classifier model.
         model_named_parameters (Iterable[Tuple[str, torch.nn.Parameter]]): model named parameters
             from ``model.named_parameters()``.
+    Notes:
+        - ``output.bias`` will be ignored if present in ``state_dict``
+        - This function will always replace the ``output.weight`` in ``state_dict``,
+            if ``output.weight != model.output.weight``.
 
     Raises:
         AssertionError: if ``state_dict`` does not contain ``output.weight``.
+        AssertionError: if ``model_named_parameters`` does not contain ``output.weight``.
+
     """
-    output_weight = [
-        (k, v) for (k, v) in model_named_parameters if k == "output.weight"
-    ]
+    output_weight = dict(model_named_parameters).get("output.weight", None)
```

Comment:
just a thought: if we have to do dict, does it create a copy? If so, would it be super slow for larger models? Not sure whats happening behind the scenes

Maybe there is some trick like model_named_parameters.args.get("output.weight", None)?

Reply:
I don't think it does a deep copy of the underlying tensor, I had a quick test. model_named_parameters returns an iterable so needs to be consumed in some way
I agree I'd like for there to be a nice way. Open to suggestions.